cl/phase1/network/services/voluntary_exit_service.go

```diff
@@ -0,0 +1,119 @@
+	if s.operationsPool.VoluntaryExitPool.Has(voluntaryExit.ValidatorIndex) {
+		return ErrIgnore
+	}
+
+	// ref: https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#voluntary-exits
+	// def process_voluntary_exit(state: BeaconState, signed_voluntary_exit: SignedVoluntaryExit) -> None:
+	state := s.syncedDataManager.HeadState()
+	if state == nil {
+		return ErrIgnore
+	}
+	val, err := state.ValidatorForValidatorIndex(int(voluntaryExit.ValidatorIndex))
+	if err != nil {
+		return ErrIgnore
+	}
+	curEpoch := s.ethClock.GetCurrentEpoch()
+
+	// Verify the validator is active
+	// assert is_active_validator(validator, get_current_epoch(state))
+	if !val.Active(curEpoch) {
+		return fmt.Errorf("validator is not active")
+	}
+
+	// Verify exit has not been initiated
+	// assert validator.exit_epoch == FAR_FUTURE_EPOCH
+	if !(val.ExitEpoch() == s.beaconCfg.FarFutureEpoch) {
+		return fmt.Errorf("verify exit has not been initiated. exitEpoch: %d, farFutureEpoch: %d", val.ExitEpoch(), s.beaconCfg.FarFutureEpoch)
+	}
+
+	// Exits must specify an epoch when they become valid; they are not valid before then
+	// assert get_current_epoch(state) >= voluntary_exit.epoch
+	if !(curEpoch >= voluntaryExit.Epoch) {
+		return fmt.Errorf("exits must specify an epoch when they become valid; they are not valid before then")
+	}
+
+	// Verify the validator has been active long enough
+	// assert get_current_epoch(state) >= validator.activation_epoch + SHARD_COMMITTEE_PERIOD
+	if !(curEpoch >= val.ActivationEpoch()+s.beaconCfg.ShardCommitteePeriod) {
+		return fmt.Errorf("verify the validator has been active long enough")
+	}
+
+	// Verify signature
+	// domain = get_domain(state, DOMAIN_VOLUNTARY_EXIT, voluntary_exit.epoch)
+	// signing_root = compute_signing_root(voluntary_exit, domain)
+	// assert bls.Verify(validator.pubkey, signing_root, signed_voluntary_exit.signature)
+	pk := val.PublicKey()
+	domainType := s.beaconCfg.DomainVoluntaryExit
+	var domain []byte
+	if state.Version() < clparams.DenebVersion {
+		domain, err = state.GetDomain(domainType, voluntaryExit.Epoch)
+	} else if state.Version() >= clparams.DenebVersion {
+		domain, err = fork.ComputeDomain(domainType[:], utils.Uint32ToBytes4(uint32(state.BeaconConfig().CapellaForkVersion)), state.GenesisValidatorsRoot())
+	}
+	if err != nil {
+		return err
+	}
+	signingRoot, err := fork.ComputeSigningRoot(voluntaryExit, domain)
+	if err != nil {
+		return err
+	}
+	if valid, err := bls.Verify(msg.Signature[:], signingRoot[:], pk[:]); err != nil {
+		return err
+	} else if !valid {
+		return errors.New("ProcessVoluntaryExit: BLS verification failed")
+	}
+
+	s.operationsPool.VoluntaryExitPool.Insert(voluntaryExit.ValidatorIndex, msg)
+
+	// Initiate exit
+	// initiate_validator_exit(state, voluntary_exit.validator_index)
+	return state.InitiateValidatorExit(voluntaryExit.ValidatorIndex)
```

Comment:
this should be `return nil`, you dont want to initialize the exit when receiving the message.